cis194/week5/tien/src/Calc.hs

```diff
@@ -0,0 +1,74 @@
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE ViewPatterns         #-}
+
+module Calc where
+
+import           ExprT
+import           Parser
+import           StackVM
+
+-- Exercise 1 --
+eval :: ExprT -> Integer
+eval (Lit a)         = a
+eval (ExprT.Add a b) = eval a + eval b
+eval (ExprT.Mul a b) = eval a * eval b
+
+-- Exercise 2 --
+evalStr :: String -> Maybe Integer
+evalStr (parseExp Lit ExprT.Add ExprT.Mul -> justExpr) = eval <$> justExpr
+
+-- Exercise 3 --
+class Expr a where
+  lit :: Integer -> a
+  add :: a -> a -> a
+  mul :: a -> a -> a
+
+instance Expr ExprT where
+  lit a = Lit a
+  add a b = ExprT.Add a b
+  mul a b = ExprT.Mul a b
+
+-- Exercise 4 --
+newtype MinMax =
+  MinMax Integer
+  deriving (Eq, Show)
+
+newtype Mod7 =
+  Mod7 Integer
+  deriving (Eq, Show)
+
+instance Expr Integer where
+  lit a = a
+  add a b = a + b
+  mul a b = a * b
+
+instance Expr Bool where
+  lit a = a > 0
+  add a b = a || b
+  mul a b = a && b
+
+instance Expr MinMax where
+  lit a = MinMax a
+  add (MinMax a) (MinMax b) = MinMax $ max a b
+  mul (MinMax a) (MinMax b) = MinMax $ min a b
+
+instance Expr Mod7 where
+  lit a = Mod7 $ mod a 7
+  add (Mod7 a) (Mod7 b) = Mod7 $ lit a + lit b
```

Comment:
This line only works because we have defined `Integer` instance off `Expr a` typeclass.
The `+` used here says `lit a` and `lit b` must be of instance `Num a`. It just so happened that `Expr Integer` is defined earlier to make `lit a + lit b` possible because `Integer` is an instance of `Num a` type class.
But in order to fulfil the `modular` requirement, that is `(3 + 5) == 1`, we need make sure `a + b` is passed to lit instead.